### ISSUE TITLE
disks with same albumartist and title grouped together

### ISSUE THREAD

**Issue by csrhodes** [LMS-Community/slimserver#45]
Hi,
I have two albums in my collection, both entitled Grands Motets, and both with album artist being Jean-Philippe Rameau.  In the out-of-the-box configuration of slimserver 7.9, the two albums are grouped together as one.
However, the two albums have different musicbrainz_album_id tags, and under normal circumstances I'd hope that would be enough to distinguish them.  I've read a bit of the schema code, and found

```
# Bug 10583 - Also check musicbrainz_id if defined.
# Can't be used in groupdiscs mode since id is unique per disc, not per set.
```

But I don't think that's right -- the musicbrainz_album_id is unique per musicbrainz release (i.e. it is the same across a multi-disc set); I have verified this by looking at other compilations in my collection.  (It's possible that this was not the case, some time ago, before the musicbrainz “Next Generation Schema” of 2011).
What's the best way to proceed?  Is there a simple way to add the musicbrainz_id checking in groupdiscs mode (and is that likely to fix my problem?)

Thanks,
Christophe

**Comment by mherger** [comment]
Are these tracks all stored in the same folder? They shouldn't be merged to the same disk if stored in different folders.

**Comment by csrhodes** [comment]
mherger notifications@github.com writes:

> Are these tracks all stored in the same folder? They shouldn't be
> merged to the same disk if stored in different folders.

No, they're in different folders.  (I'm happy to help debug this on my
collection; let me know what you want me to do.)

Christophe

**Comment by mherger** [comment]
Please feel free to send me your library.db (compressed - michael ät slim devices dot com). Is this a regression in 7.9? Does it not show up in earlier versions?

**Comment by csrhodes** [comment]
mherger notifications@github.com writes:

> Please feel free to send me your library.db (compressed).

Thanks.  I've put it up (temporarily) at
http://www.doc.gold.ac.uk/~mas01cr/tmp/library.db.bz2; the album in
question is album 165.

> Is this a regression in 7.9? Does it not show up in earlier versions?

Unfortunately, I'm not sure.  I'm pretty sure my previous installation
didn't suffer from this problem, but that was possibly with different
server settings (I'm not sure about the various artists setting, for
example), and that machine is now switched off.  I can turn it on at
some point if that would help, but not terribly easily.

Best,

Christophe

**Comment by mherger** [comment]
Thanks for the file. I think the issue is described a little further down the code than the comment you pasted:

```
        # Bug 10583 - If we had the MUSICBRAINZ_ALBUM_ID in the tracks table,
        # we could join on it here ...
        # TODO: Join on MUSICBRAINZ_ALBUM_ID if it ever makes it into the tracks table.

        # Join on tracks with the same basename to determine a unique album.
        # Bug 10583 - Only try to aggregate from basename
        # if no MUSICBRAINZ_ALBUM_ID and no DISC and no DISCC available.
        # Bug 11780 - Need to handle groupdiscs mode differently; would leave out
        # basename check if MB Album Id given and thus merge different albums
        # of the same name into one.
```

Which I read like: MUSICBRAINZ_ALBUM_ID isn't supported yet, but we don't check on the file path either if it is available. Or the MUSICBRAINZ_ALBUM_ID values in your various files are identical?

Could you please run a full wipe & rescan on the Jean-Philippe Rameau folder only, with all scanner log options set to debug, then provide the scanner.log file?

**Comment by csrhodes** [comment]
mherger notifications@github.com writes:

> Which I read like: MUSICBRAINZ_ALBUM_ID isn't supported yet, but we
> don't check on the file path either if it is available. Or the
> MUSICBRAINZ_ALBUM_ID values in your various files are identical?

The musicbrainz_album_id values in my two distinct albums are distinct
:-).  But yes, I think the comments there are pertinent -- but I think
they might be addressing a five-year-old problem; in the old days
(before 2011) musicbrainz had one _album_id per disc; now it is one
_album_id per release (so multi-disc releases have the same
musicbrainz_album_id).  So the code for the group discs case could be
reused?  (Sorry, I'm away from the machine and can't really look at the
code in detail right now; I hope the idea is clear even if I've got the
terminology wrong).  Of course there might be users who have tags from
the old Musicbrainz schema :-/

> Could you please run a full wipe & rescan on the Jean-Philippe Rameau
> folder only, with all scanner log options set to debug, then provide
> the scanner.log file?

Is there a way of selectively forgetting about files or folders?  Either
from the command-line or the web interface?  I guess the intent here is
to watch the scan of just the two albums in question -- I have other
Rameau discs which would probably just muddy the waters.

**Comment by csrhodes** [comment]
mherger notifications@github.com writes:

> Could you please run a full wipe & rescan on the Jean-Philippe Rameau
> folder only, with all scanner log options set to debug, then provide
> the scanner.log file?

I'm not sure I've done this right -- but I've attached the scanner.log
from scanning just the Rameau directory (which has a few other discs as
well as the two in question).  I hope it helps.

Best,

Christophe

**Comment by mherger** [comment]
Where did you attach them?

**Comment by csrhodes** [comment]
mherger notifications@github.com writes:

> Where did you attach them?

Bother.  Sorry, I trusted that github would do the right thing with
e-mailed attachments.  I've uploaded the scanner log to
http://www.doc.gold.ac.uk/~mas01cr/tmp/scanner.log.gz.  (I'm sure I've
done it wrong, or at least the log doesn't look very helpful to me; let
me know what I should have done...)

Best,

Christophe

**Comment by mherger** [comment]
> Bother. Sorry, I trusted that github would do the right thing with
> e-mailed attachments. I've uploaded the scanner log to
> http://www.doc.gold.ac.uk/~mas01cr/tmp/scanner.log.gz. (I'm sure I've
> done it wrong, or at least the log doesn't look very helpful to me; let
> me know what I should have done...)

That file seems to be missing output from scan.import, only has 
scan.scanner enabled. Did you set _all_ scan.\* options to debug before 
you run the server?

## 

Michael Herger

**Comment by csrhodes** [comment]
mherger notifications@github.com writes:

> That file seems to be missing output from scan.import, only has 
> scan.scanner enabled. Did you set _all_ scan.\* options to debug before 
> you run the server?

I thought I did.  I've rerun the scan, resulting in
http://www.doc.gold.ac.uk/~mas01cr/tmp/scanner.log (uncompressed this
time; it's only 300k) after restarting the server: the
prefs/log.conf file looks like
http://www.doc.gold.ac.uk/~mas01cr/tmp/log.conf.  (I don't think the
output is materially different, unfortunately).

Christophe

**Comment by mherger** [comment]
Heh... took me a while to figure this simple one out: you'll need to enable database.info, rather than the scan.\* options. Don't know why this was decided this way. Could you please try again? Thanks!

**Comment by csrhodes** [comment]
mherger notifications@github.com writes:

> Heh... took me a while to figure this simple one out: you'll need to
> enable database.info, rather than the scan.\* options. Don't know why
> this was decided this way. Could you please try again? Thanks!

Aha, that did something.  (I found I had to do a full server restart:
just changing the logging parameters in the web interface didn't change
the logging, for whatever reason).  However, once I did that I got much
larger output; I've uploaded it to
http://www.doc.gold.ac.uk/~mas01cr/tmp/scanner.log.gz.

Thanks,

Christophe

**Comment by mherger** [comment]
Thanks - I'll look into this asap.

**Comment by mherger** [comment]
Chrisophe - I've looked in to this issue a bit more. Unfortunately you suggest the exact opposite of what a user said when above comment was made wrt. bug 10583: he wrote:

_"The only caveat to think about might be if they enable grouping of disc sets in which case MusicBrainz will have a different MUSICBRAINZ_ALBUM_ID for EACH DISC in the set."_

see http://bugs.slimdevices.com/show_bug.cgi?id=10583#c8

Now I'm hesitating to change code which has been in place for 5+ years before somebody complained about it... The risk to break more than I'd fix seems rather high to me. Do you have additional information about how musicbrainz is dealing with disc sets etc.?

**Comment by csrhodes** [comment]
Hi,

That user's comment was made in 2009.  In 2011, Musicbrainz changed their schema to the "Next Generation Schema" – essentially a complete rewrite of the Musicbrainz data model.  In particular, pre-NGS there was no concept of a "release" as distinct from a single disc (you can see a discussion about this in the release notes for the new schema at the time: https://wiki.musicbrainz.org/Server_Release_Notes/20110516).  In the Next Generation Schema, the `musicbrainz_albumid` tag corresponds to a Release (see http://picard.musicbrainz.org/docs/mappings/), which is (roughly) "what the consumer buys" -- i.e. if it's a multi-disc album, there's one `musicbrainz_albumid` which is consistent across all the discs (see http://musicbrainz.org/doc/Release).

The overall database schema as it exists now is http://wiki.musicbrainz.org/-/images/5/52/ngs.png -- for the purposes of this issue, the relevant point is that a release, corresponding to `musicbrainz_albumid` (in the south-west quadrant) has one or more mediums, corresponding to physical discs.

I hope that helps.  It probably can't really help with evaluating the risks of breaking more than fixing; that depends on what data the world out there has, and how able they are to retag things (if they've kept tags from pre-2011).  But as far as I can tell, the user's comment in that bug was true in 2009 and has been untrue since 2011...

Cheers,

Christophe

**Comment by mherger** [comment]
Now that's a great comment with a lot of useful information I needed. I'll review this. Thanks!

**Comment by mherger** [comment]
@csrhodes - I get from your github profile you'd be able to apply a patch to test this change? Or run LMS from a git clone?

**Comment by csrhodes** [comment]
Yes -- I can run from either a patch or a branch; whatever's easiest.  (It will probably take me a little while to get to it; this week is busy for me...)

**Comment by mherger** [comment]
Could it be as simple as this:

``` diff
diff --git a/Slim/Schema.pm b/Slim/Schema.pm
index 77ca1e0..954eb60 100644
--- a/Slim/Schema.pm
+++ b/Slim/Schema.pm
@@ -962,8 +962,9 @@ sub _createOrUpdateAlbum {
                my $checkDisc = 0;

                # Bug 10583 - Revert disc 1/1 change. Use MB Album Id in addition (unique id per disc, not per set!)
+               # XXX - as of 2011 Musicbrainz New Generation Schema treats the set as an album, not the disc (as before).
                if (!$prefs->get('groupdiscs') && 
-                       (($disc && $discc) || ($disc && !$discc) || $brainzId)) {
+                       (($disc && $discc) || ($disc && !$discc))) {

                        $checkDisc = 1;
                }
@@ -1093,9 +1094,8 @@ sub _createOrUpdateAlbum {
                        # basename check if MB Album Id given and thus merge different albums
                        # of the same name into one.
                        if (
-                               # In checkDisc mode, try "same folder" only if none of MUSICBRAINZ_ALBUM_ID,
-                               # DISC and DISCC are known.
-                               ($checkDisc && !defined $brainzId && !defined $disc && !defined $discc) ||
+                               # In checkDisc mode, try "same folder" only if none of DISC and DISCC are known.
+                               ($checkDisc && !defined $disc && !defined $discc) ||
                                # When not checking discs (i.e., "Group Discs" mode), try "same folder"
                                # as a last resort if both DISC and DISCC are unknown.
                                (!$checkDisc && !defined $disc && !defined $discc)
```

**Comment by csrhodes** [comment]
sadly not; I've rerun the scan and uploaded the scanner log at http://www.doc.gold.ac.uk/~mas01cr/tmp/scanner-20150321.log.gz.

I noodled about a bit in Schema.pm, and the patch at http://www.doc.gold.ac.uk/~mas01cr/tmp/slim.diff works for me (there might be tab/space breakage, sorry -- apply with -l).  I think it's the right thing given the New Generation Schema, at least judging by the comment, but I'm not sure.

**Comment by mherger** [comment]
Fixed by pull request #87
